### ISSUE TITLE
Clarify the docs and setting for lsp-file-watch-ignored

### ISSUE THREAD

**Issue by madscientist** [emacs-lsp/lsp-mode#828]
**Describe the bug**
After examining the code it appears that the values in `lsp-file-watch-ignored` are actually regexs not plain strings.  However neither the documentation nor the default values of this list reflect that.

It would be helpful if this was noted in the doc string for the variable, and also the default values appear to be incorrect; for example they contain `.git` which, if I'm understanding things correctly, would match something like `fgit` as well.  Shouldn't the default values be escaped, like `"\\.git"`?

Finally, in the code we loop over `(directory-files dir t)`; the `t` here means (as I understand it) we're matching against the full path of each directory.  So, shouldn't the above regex values also start with `/` to specify they must be at the beginning of a directory (do we have to use `[/\\]` to support Windows paths?), else wouldn't they also match things like `foo.git`?  And shouldn't they end with `$` so they don't match `.gitsubdir`?

Finally, I think it would be helpful to mark `lsp-file-watch-ignored` as `safe-local-variable` using something like:
```
(put 'lsp-file-watch-ignored 'safe-local-variable 'listp)
```
(I dunno, maybe there's a more selective predicate function that matches only lists of strings).

There may be other lsp variables which would be useful to so mark as well, so people can customize them on a per-project basis with `.dir-locals.el`.

**Comment by yyoncho** [comment]
> So, shouldn't the above regex values also start with / to specify they must be at the beginning of a directory (do we have to use [/\\] to support Windows paths?), else wouldn't they also match things like foo.git? And shouldn't they end with $ so they don't match .gitsubdir?

Yes, I think that we should do what you are proposing here with questions. Do you wanna work on PR?

**Comment by madscientist** [comment]
OK can do

**Comment by madscientist** [comment]
Question: is there another/better way to see which watches are registered, than looking at the `*lsp-log*` buffer?  That's simple enough, just wondering if there's a more exact way.

**Comment by yyoncho** [comment]
This gives the dirs with watches `(ht-keys (lsp-session-watches (lsp-session)))`

**Comment by madscientist** [comment]
Thanks; that appears to only give the root directory, though, not all the recursive directories...?  When I eval that in a buffer in my workspace I get a list with a single element but in `*lsp-log*` I can see lots of watched subdirectories.

**Comment by yyoncho** [comment]
The subdirectories are available via ht-vals.

**Comment by madscientist** [comment]
Got it thx.

**Comment by madscientist** [comment]
While working on this, I noticed that it seems `lsp-file-watch-ignored` only applies to directories which exist when the project is started up (e.g., when the lsp hook is called for the first time in a project).  If I go in after that and make new directories which match any of the ignored regexps, that new directory is watched which is unexpected.

I don't know if this is a known issue?

**Comment by yyoncho** [comment]
@madscientist see https://github.com/emacs-lsp/lsp-mode/issues/805 

User @aureumleonis wanted to work on it but I guess you could fix it too if it is blocking you.

**Comment by madscientist** [comment]
I wasn't planning on working on it.  I'll leave a comment there with what I'm seeing which looks slightly different (but could well be the same issue).

Can people confirm that the `make test` at master HEAD is 100% successful?  On my system, even without any changes to the `lsp-mode` code, I see:
```
[lsp-mode|master]$ make test
cask build
Compiling ...

cask exec ert-runner
...................Test lsp-file-notifications-test backtrace:

  signal(ert-test-failed (((should (equal lsp--test-events `(((workspa
  ert-fail(((should (equal lsp--test-events `(((workspace-1) "workspac
  (if (unwind-protect (setq value-197 (apply fn-195 args-196)) (setq f
  (let (form-description-199) (if (unwind-protect (setq value-197 (app
  (let ((value-197 'ert-form-evaluation-aborted-198)) (let (form-descr
  (let* ((fn-195 (function equal)) (args-196 (condition-case err (let
  (let* ((root (make-temp-file "test-root-directory" t)) (lsp--cur-wor
  (closure (t) nil (let* ((root (make-temp-file "test-root-directory"
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name lsp-file-notifications-test :document
  ert-run-or-rerun-test(#s(ert--stats :selector (and t) :tests [#s(ert
  ert-run-tests((and t) (lambda (event-type &rest event-args) (cond ((
  ert-runner/run-tests-batch((and t))
  (let ((stats (ert-runner/run-tests-batch selector))) (kill-emacs (if
  ert-runner/run-tests-batch-and-exit((and t))
  (if ert-runner-verbose (ert-runner/run-tests-batch-and-exit ert-runn
  (let ((test-files (ert-runner--test-files tests)) (test-helper (f-ex
  ert-runner/run()
  apply(ert-runner/run nil)
  commander--handle-command(nil)
  commander-parse(nil)
  (if commander-parsing-done nil (commander-parse (or commander-args (
  eval-buffer(#<buffer  *load*> nil "/home/ms/src/lsp-mode/.cask/2
  load-with-code-conversion("/home/ms/src/lsp-mode/.cask/26.1/elpa
  load("/home/ms/src/lsp-mode/.cask/26.1/elpa/ert-runner-20180831.
  command-line-1(("-scriptload" "/home/ms/src/lsp-mode/.cask/26.1/
  command-line()
  normal-top-level()

Test lsp-file-notifications-test condition:

    (ert-test-failed
     ((should
       (equal lsp--test-events `...))
      :form
      (equal nil
             ((... "workspace/didChangeWatchedFiles" ...)
              (... "workspace/didChangeWatchedFiles" ...)))
      :value nil :explanation
      (different-types nil
                       ((... "workspace/didChangeWatchedFiles" ...)
                        (... "workspace/didChangeWatchedFiles" ...)))))

FTest lsp-file-watch--adding-watches backtrace:

  signal(ert-test-failed (((should (equal expected-events events)) :fo
  ert-fail(((should (equal expected-events events)) :form (equal ((cha
  (if (unwind-protect (setq value-187 (apply fn-185 args-186)) (setq f
  (let (form-description-189) (if (unwind-protect (setq value-187 (app
  (let ((value-187 'ert-form-evaluation-aborted-188)) (let (form-descr
  (let* ((fn-185 (function equal)) (args-186 (condition-case err (let
  (let* ((temp-directory (make-temp-file "test-dir" t)) (nested-dir (f
  (closure (t) nil (let* ((temp-directory (make-temp-file "test-dir" t
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name lsp-file-watch--adding-watches :docum
  ert-run-or-rerun-test(#s(ert--stats :selector (and t) :tests [#s(ert
  ert-run-tests((and t) (lambda (event-type &rest event-args) (cond ((
  ert-runner/run-tests-batch((and t))
  (let ((stats (ert-runner/run-tests-batch selector))) (kill-emacs (if
  ert-runner/run-tests-batch-and-exit((and t))
  (if ert-runner-verbose (ert-runner/run-tests-batch-and-exit ert-runn
  (let ((test-files (ert-runner--test-files tests)) (test-helper (f-ex
  ert-runner/run()
  apply(ert-runner/run nil)
  commander--handle-command(nil)
  commander-parse(nil)
  (if commander-parsing-done nil (commander-parse (or commander-args (
  eval-buffer(#<buffer  *load*> nil "/home/ms/src/lsp-mode/.cask/2
  load-with-code-conversion("/home/ms/src/lsp-mode/.cask/26.1/elpa
  load("/home/ms/src/lsp-mode/.cask/26.1/elpa/ert-runner-20180831.
  command-line-1(("-scriptload" "/home/ms/src/lsp-mode/.cask/26.1/
  command-line()
  normal-top-level()

Test lsp-file-watch--adding-watches condition:

    (ert-test-failed
     ((should
       (equal expected-events events))
      :form
      (equal
       ((changed "/tmp/test-dirIVSa2o/nested/file.ext")
        (created "/tmp/test-dirIVSa2o/nested/file.ext"))
       nil)
      :value nil :explanation
      (different-types
       ((changed "/tmp/test-dirIVSa2o/nested/file.ext")
        (created "/tmp/test-dirIVSa2o/nested/file.ext"))
       nil)))

F...Test lsp-file-watch--recursive backtrace:

  signal(ert-test-failed (((should (equal events expected-events)) :fo
  ert-fail(((should (equal events expected-events)) :form (equal nil (
  (if (unwind-protect (setq value-112 (apply fn-110 args-111)) (setq f
  (let (form-description-114) (if (unwind-protect (setq value-112 (app
  (let ((value-112 'ert-form-evaluation-aborted-113)) (let (form-descr
  (let* ((fn-110 (function equal)) (args-111 (condition-case err (let
  (let* ((temp-directory (make-temp-file "test-dir" t)) (matching-file
  (closure (t) nil (let* ((temp-directory (make-temp-file "test-dir" t
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name lsp-file-watch--recursive :documentat
  ert-run-or-rerun-test(#s(ert--stats :selector (and t) :tests [#s(ert
  ert-run-tests((and t) (lambda (event-type &rest event-args) (cond ((
  ert-runner/run-tests-batch((and t))
  (let ((stats (ert-runner/run-tests-batch selector))) (kill-emacs (if
  ert-runner/run-tests-batch-and-exit((and t))
  (if ert-runner-verbose (ert-runner/run-tests-batch-and-exit ert-runn
  (let ((test-files (ert-runner--test-files tests)) (test-helper (f-ex
  ert-runner/run()
  apply(ert-runner/run nil)
  commander--handle-command(nil)
  commander-parse(nil)
  (if commander-parsing-done nil (commander-parse (or commander-args (
  eval-buffer(#<buffer  *load*> nil "/home/ms/src/lsp-mode/.cask/2
  load-with-code-conversion("/home/ms/src/lsp-mode/.cask/26.1/elpa
  load("/home/ms/src/lsp-mode/.cask/26.1/elpa/ert-runner-20180831.
  command-line-1(("-scriptload" "/home/ms/src/lsp-mode/.cask/26.1/
  command-line()
  normal-top-level()

Test lsp-file-watch--recursive condition:

    (ert-test-failed
     ((should
       (equal events expected-events))
      :form
      (equal nil
             ((changed "/tmp/test-dirVpk85Q/file.ext")
              (created "/tmp/test-dirVpk85Q/file.ext")))
      :value nil :explanation
      (different-types nil
                       ((changed "/tmp/test-dirVpk85Q/file.ext")
                        (created "/tmp/test-dirVpk85Q/file.ext")))))

FTest lsp-file-watch--relative-path-glob-patterns backtrace:

  signal(ert-test-failed (((should (equal expected-events events)) :fo
  ert-fail(((should (equal expected-events events)) :form (equal ((cha
  (if (unwind-protect (setq value-157 (apply fn-155 args-156)) (setq f
  (let (form-description-159) (if (unwind-protect (setq value-157 (app
  (let ((value-157 'ert-form-evaluation-aborted-158)) (let (form-descr
  (let* ((fn-155 (function equal)) (args-156 (condition-case err (let
  (let* ((temp-directory (file-name-as-directory (concat temporary-fil
  (closure (t) nil (let* ((temp-directory (file-name-as-directory (con
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name lsp-file-watch--relative-path-glob-pa
  ert-run-or-rerun-test(#s(ert--stats :selector (and t) :tests [#s(ert
  ert-run-tests((and t) (lambda (event-type &rest event-args) (cond ((
  ert-runner/run-tests-batch((and t))
  (let ((stats (ert-runner/run-tests-batch selector))) (kill-emacs (if
  ert-runner/run-tests-batch-and-exit((and t))
  (if ert-runner-verbose (ert-runner/run-tests-batch-and-exit ert-runn
  (let ((test-files (ert-runner--test-files tests)) (test-helper (f-ex
  ert-runner/run()
  apply(ert-runner/run nil)
  commander--handle-command(nil)
  commander-parse(nil)
  (if commander-parsing-done nil (commander-parse (or commander-args (
  eval-buffer(#<buffer  *load*> nil "/home/ms/src/lsp-mode/.cask/2
  load-with-code-conversion("/home/ms/src/lsp-mode/.cask/26.1/elpa
  load("/home/ms/src/lsp-mode/.cask/26.1/elpa/ert-runner-20180831.
  command-line-1(("-scriptload" "/home/ms/src/lsp-mode/.cask/26.1/
  command-line()
  normal-top-level()

Test lsp-file-watch--relative-path-glob-patterns condition:

    (ert-test-failed
     ((should
       (equal expected-events events))
      :form
      (equal
       ((changed "/tmp/common-test-dir/file.ext")
        (created "/tmp/common-test-dir/file.ext"))
       nil)
      :value nil :explanation
      (different-types
       ((changed "/tmp/common-test-dir/file.ext")
        (created "/tmp/common-test-dir/file.ext"))
       nil)))

F...........

Ran 37 tests in 6.999 seconds
4 unexpected results:
   FAILED  lsp-file-notifications-test
   FAILED  lsp-file-watch--adding-watches
   FAILED  lsp-file-watch--recursive
   FAILED  lsp-file-watch--relative-path-glob-patterns
make: *** [Makefile:7: test] Error 1
```

**Comment by yyoncho** [comment]
The file notification tests does not work unless you are using gui since they require sit-for function.

**Comment by madscientist** [comment]
Oh OK thanks.  Since these seem to be the tests most relevant to the changes I'm making, are there instructions somewhere for running them in an interactive Emacs session instead of batch mode?  I'm not that familiar with ert and ert-runner.

**Comment by yyoncho** [comment]
you may use `ert-run-tests-interactively` or just eval the block in the test.

**Comment by madscientist** [comment]
Unfortunately I'm too ignorant to know what to do with that information.  I tried running `cask emacs -Q -l lsp` then the above command, but it didn't find any tests to run so obviously there's something else that has to happen before these commands will work and I can't discover it from any docs etc. I've managed to find so far.

I guess I'll just submit the PR and someone else can run the tests instead :).
